src/Common/src/TypeSystem/TypesDebugInfoWriter/UserDefinedTypeDescriptor.cs

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Reflection.Metadata;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.TypeSystem.TypesDebugInfo
+{
+    public class UserDefinedTypeDescriptor
+    {
+        public UserDefinedTypeDescriptor(ITypesDebugInfoWriter objectWriter)
+        {
+            _objectWriter = objectWriter;
+        }
+        public uint GetTypeIndex(TypeDesc type)
+        {
+            uint variableTypeIndex = 0;
+            if (!_knownTypes.TryGetValue(type, out variableTypeIndex))
+            {
+                variableTypeIndex = GetNewTypeIndex(type);
+            }
+            return variableTypeIndex;          
+        }
+
+        private uint GetNewTypeIndex(TypeDesc type)
+        {
+            switch (type.Category)
```

Comment:
Don't switch on `Category` if you can help it. It has a lot of things in it that you probably won't catch. E.g. `TypeFlags.Nullable` and `TypeFlags.Interface` should be handled the same as `ValueType` and `Class` here. We might add other flags later.

Just use `type.IsEnum` and `type.IsDefType`. Those will always be correct.